From b838b68d69982d679dd4c131456f4b9ac9408e75 Mon Sep 17 00:00:00 2001
From: turbo-boo <59400127+turbo-boo@users.noreply.github.com>
Date: Sun, 4 Oct 2026 12:52:32 +0900
Subject: [PATCH] display: Add Xiaomi CWB sensor integration

Notify Xiaomi's CIT sensor service through libclient2slpi.notifier on
built-in display frame updates when vendor.display.enable_cwb_notify_sensor
is set, so the under-display light sensor can trigger CWB sampling.

Change-Id: I67dd5db1eb0d1f9975190ef0efb48be96fd4a18a
---
 Android.bp                        | 25 +++++++++++++++++++++++++
 include/display_properties.h      |  1 +
 sdmclient/Android.bp              |  5 ++++-
 sdmclient/client_rgb_notifier.h   | 20 ++++++++++++++++++++
 sdmclient/sdm_display.cpp         | 12 ++++++++++++
 sdmclient/sdm_display.h           |  7 +++++++
 sdmclient/sdm_display_builtin.cpp |  7 +++++++
 7 files changed, 76 insertions(+), 1 deletion(-)
 create mode 100644 sdmclient/client_rgb_notifier.h

diff --git a/Android.bp b/Android.bp
index e85c6b9..d7ed948 100644
--- a/Android.bp
+++ b/Android.bp
@@ -1,3 +1,28 @@
+soong_config_module_type {
+    name: "xiaomi_display_cc_defaults",
+    module_type: "cc_defaults",
+    config_namespace: "xiaomi_display",
+    bool_variables: [
+        "cwb_notify_sensor",
+    ],
+    properties: [
+        "cflags",
+        "shared_libs",
+    ],
+}
+
+xiaomi_display_cc_defaults {
+    name: "xiaomi_cwb_notify_defaults",
+    soong_config_variables: {
+        cwb_notify_sensor: {
+            cflags: ["-DXIAOMI_CWB_NOTIFY_SENSOR"],
+            shared_libs: [
+                "//vendor/xiaomi/sm8750-common:libclient2slpi.notifier",
+            ],
+        },
+    },
+}
+
 cc_library_headers {
     name: "display_debug_headers",
     vendor: true,
diff --git a/include/display_properties.h b/include/display_properties.h
index b31f701..bbb3fe8 100644
--- a/include/display_properties.h
+++ b/include/display_properties.h
@@ -254,6 +254,7 @@
 #define DISABLE_CWB_OPTIMIZATION             DISPLAY_PROP("disable_cwb_optimization")
 #define FORCE_REFRESH_TO_PROCESS_CWB         DISPLAY_PROP("force_refresh_to_process_cwb")
 #define ENABLE_CLIENT_CONTROL_CWB_REFRESH    DISPLAY_PROP("enable_client_control_cwb_refresh")
+#define ENABLE_CWB_NOTIFY_SENSOR             DISPLAY_PROP("enable_cwb_notify_sensor")
 #define ENABLE_ASYNC_POWER_OFF_WAIT          DISPLAY_PROP("enable_async_power_off_wait")
 #define TRANSIENT_FPS_CYCLE_COUNT            DISPLAY_PROP("transient_fps_cycle_count")
 #define CORE_ID_MASK                         DISPLAY_PROP("core_id_mask")
diff --git a/sdmclient/Android.bp b/sdmclient/Android.bp
index 69865e8..9e07ee5 100644
--- a/sdmclient/Android.bp
+++ b/sdmclient/Android.bp
@@ -1,6 +1,9 @@
 cc_library_shared {
     name: "libsdmclient",
-    defaults: ["qtidisplay_defaults"],
+    defaults: [
+        "qtidisplay_defaults",
+        "xiaomi_cwb_notify_defaults",
+    ],
     sanitize: {
         integer_overflow: true,
     },
diff --git a/sdmclient/client_rgb_notifier.h b/sdmclient/client_rgb_notifier.h
new file mode 100644
index 0000000..0c7149d
--- /dev/null
+++ b/sdmclient/client_rgb_notifier.h
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2026 The LineageOS Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __CLIENT_RGB_NOTIFIER_H__
+#define __CLIENT_RGB_NOTIFIER_H__
+
+#include <utils/RefBase.h>
+
+// Implemented by Xiaomi's libclient2slpi.notifier.so.
+class clientRGBNotifier : public android::RefBase {
+ public:
+  clientRGBNotifier();
+  virtual ~clientRGBNotifier();
+
+  void SendFrameUpdateSingnal(bool is_forced);
+};
+
+#endif  // __CLIENT_RGB_NOTIFIER_H__
diff --git a/sdmclient/sdm_display.cpp b/sdmclient/sdm_display.cpp
index bd34194..8da2254 100644
--- a/sdmclient/sdm_display.cpp
+++ b/sdmclient/sdm_display.cpp
@@ -582,6 +582,18 @@ DisplayError SDMDisplay::Init() {
     swap_interval_zero_ = true;
   }
 
+#ifdef XIAOMI_CWB_NOTIFY_SENSOR
+  if (type_ == kBuiltIn) {
+    int enable_cwb_notify_sensor = 0;
+    SDMDebugHandler::Get()->GetProperty(ENABLE_CWB_NOTIFY_SENSOR,
+                                        &enable_cwb_notify_sensor);
+    enable_cwb_notify_sensor_ = enable_cwb_notify_sensor;
+    if (enable_cwb_notify_sensor_) {
+      client_rgb_notifier_ = new clientRGBNotifier();
+    }
+  }
+#endif
+
   idle_active_ms_ = SDMDebugHandler::GetIdleTimeoutMs();
 
   client_target_ = new SDMLayer(id_, buffer_allocator_);
diff --git a/sdmclient/sdm_display.h b/sdmclient/sdm_display.h
index 2a01d3f..391292a 100644
--- a/sdmclient/sdm_display.h
+++ b/sdmclient/sdm_display.h
@@ -50,6 +50,9 @@
 #include <vector>
 #include <climits>
 
+#ifdef XIAOMI_CWB_NOTIFY_SENSOR
+#include "client_rgb_notifier.h"
+#endif
 #include "sdm_compositor_callbacks.h"
 #include "sdm_layer_builder.h"
 
@@ -647,6 +650,10 @@ public:
   int disable_hdr_handling_ = 0;  // disables HDR handling.
   int disable_sdr_histogram_ = 0; // disables handling of SDR histogram data.
   bool pending_commit_ = false;
+#ifdef XIAOMI_CWB_NOTIFY_SENSOR
+  android::sp<clientRGBNotifier> client_rgb_notifier_;
+  bool enable_cwb_notify_sensor_ = false;
+#endif
   bool is_cmd_mode_ = false;
   bool partial_update_enabled_ = false;
   bool skip_commit_ = false;
diff --git a/sdmclient/sdm_display_builtin.cpp b/sdmclient/sdm_display_builtin.cpp
index f17b60d..59b3fad 100644
--- a/sdmclient/sdm_display_builtin.cpp
+++ b/sdmclient/sdm_display_builtin.cpp
@@ -1585,6 +1585,13 @@ DisplayError SDMDisplayBuiltIn::CommitOrPrepare(
     HandleLargeCompositionHint(!needs_hint);
   }
 
+#ifdef XIAOMI_CWB_NOTIFY_SENSOR
+  if (current_power_mode_ == SDMPowerMode::POWER_MODE_ON &&
+      enable_cwb_notify_sensor_ && client_rgb_notifier_ != nullptr) {
+    client_rgb_notifier_->SendFrameUpdateSingnal(is_primary_);
+  }
+#endif
+
   prepare_phase_ = false;
 
   // Need a commit call to flush the dsi dynamic clock
-- 
2.43.0

